docs/ProgrammingConcerns.md

```diff
@@ -0,0 +1,170 @@
+
+### Why not just just replace the system allocator?
+JEMalloc and similar allocators allow your application to replace the system allocator. This is something that FEX used to do, but this has since been
+removed because 32-bit thunks need to allocate memory in the lower 4GB.
+
+## APIs which allocate memory that FEX needs to avoid
+Most C++ APIs allow you to replace their allocators, but some don't and we need to avoid those APIs. If FEX uses them then the 32-bit application
+running might run out of memory. This isn't an all encompassing list and we will add to it as our CI captures more problems.
+
+### `get_nprocs_conf`
+Use `FEXCore::CPUInfo::CalculateNumberOfCPUs` instead.
+
+### `getcwd`
+Don't use getcwd with a nullptr buffer. It will allocate memory behind our back and return a pointer that needs a free.
+
+### `strerror`
+This allocates and frees memory based on locale! Even with C local it'll attempt to free(0).
+FEX-Emu should avoid using this function and instead just return the number.
+If necessary FEX will provide its own routine for getting this string back.
+
+### `std::make_unique`
+Use `fextl::make_unique` instead.
+
+### `std::unique_ptr`
+Use `fextl::unique_ptr` instead.
+
+### `std::filesystem`
+This namespace is /highly/ likely to allocate memory behind our back.
+FEX should avoid using this API as much as possible.
+
+#### std::filesystem::path
+#### std::filesystem::path::is_relative
+Use `FHU::Filesystem::IsRelative` instead.
+
+#### std::filesystem::absolute
+Always allocates memory.
+Use `realpath` instead.
+
+#### std::filesystem::exists
+Creates a std::filesystem::path when passing in to it.
+Use `FHU::Filesystem::Exists` instead.
+
+#### std::filesystem::canonical
+Always allocates memory.
+Use `realpath` instead.
+
+#### std::filesystem::path::lexically_normal
+Use `FHU::Filesystem::LexicallyNormal` instead.
+
+#### std::filesystem::create_directory
+#### std::filesystem::create_directories
+Creates a std::filesystem::path when passing in to it.
+Use `FHU::Filesystem::CreateDirectory` and `FHU::Filesystem::CreateDirectories` instead.
+
+#### std::filesystem::path::parent_path
+Use `FHU::Filesystem::ParentPath` instead.
+
+#### std::filesystem::path::filename
+Use `FHU::Filesystem::GetFilename` instead.
+
+#### std::filesystem::copy_file
+Use `FHU::Filesystem::CopyFile` instead.
+
+### `std::fstream`
+This API always allocates memory and should be avoided.
+Use a combination of open and fextl::string APIs instead of fstream.
+
+### `std::string`
+Use `fextl::string` instead.
+
+#### std::to_string
+Use `fextl::fmt::format` instead.
+
+### APIs that FEX doesn't have a replacement for
+Don't use any of these APIs in FEXLoader/FEXInterpreter. Shoutout to
+[this](https://stackoverflow.com/questions/43056338/standard-library-facilities-which-allocate-but-dont-use-an-allocator) StackOverflow post for this
+huge list.
+
+#### `std::any`
+#### `std::function`
+Ensure the std::function doesn't allocate memory.
```

Comment:
What actionable advice is this intended to give? Should clarify if we implicitly rely on a small-function optimization here, and (if so) what size this commonly limits us to with the standard libraries we compile against.

Reply:
Changed the comment around this stating how it is actually impossible to determine if memory is getting allocated or not.